### DISCUSSION TITLE
Nonce or hash for inline styles (for Content-Security-Policy)

### DISCUSSION THREAD

**Discussion by sophiekoonin** [vercel/next.js#18451] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.
Recent versions of Next are using inline styles, which break our apps because we block `style-src: unsafe-inline` in our `Content-Security-Policy` header. In fact, the default behaviour of any CSP is to block unsafe inline styles. While the risk may be less than with unsafe inline scripts, there is still a risk: see [this StackOverflow answer](https://stackoverflow.com/questions/30653698/csp-style-src-unsafe-inline-is-it-worth-it). 

We have extremely high security requirements at my company so it's not really a matter of easily being able to disable `unsafe-inline`, and I'm aware that the future intention is to move more of the styling inline. If we can't upgrade Next because of this, we'll either be stuck on an old version and potentially vulnerable that way, or we'll have to consider alternatives (which I really don't want to do!).

## Describe the solution you'd like
Inline styles should have a [hash or nonce](https://content-security-policy.com/examples/allow-inline-style/) which should be exposed as a global variable that we can inject into our CSP. A nonce is probably the easiest way forward. It will have to change on every request, so we'd need to have a way of generating this on the server side.

I can see that the `Head` component in `next/document` accepts a `nonce` prop but this doesn't appear to apply to Next-generated inline styles. 

## Describe alternatives you've considered
I'm not sure there are any real alternatives here given the intention to move most of the critical styling inline.

**Comment by bthorben** [comment]
I am facing exactly the described issue. While this seems to be hard to do with next.js, it was no problem to achieve with our, much older, custom webpack setup.

**Reply by sophiekoonin** [reply]
This has been solved now - the `nonce` you pass in to `Head` will apply to the inline styles. You can set a custom CSP header on the `res` object in `_document.js` as follows: 

```js
export default class extends Document {
  static async getInitialProps(ctx) {
    const initialProps = await Document.getInitialProps(ctx)
    const nonce = generateNonce() // however you do this is up to you
    const csp = getCSP(nonce) // we have a function that returns a string incorporating the nonce
    const res = ctx?.res
    if (res != null) {
      res.setHeader('Content-Security-Policy', csp)
    }
    return {
      ...initialProps,
      nonce,
    }
  }

  render() {
    const { nonce } = this.props

    return (
      <Html lang="en" dir="ltr">
        <Head nonce={nonce}>

etc
```

**Reply by alex-r89** [reply]
Hey @sophiekoonin thanks for sharing this!

I'm still noticing a CSP error though, specifically that both inline script's and inline styles are failing. Is passing the `nonce` into the `Head` enough in your case, or do you need to also pass it into other elements? My CSP is as follows which seems to be failing:

```js
function getCsp(nonce) {
  let csp = ``
  csp += `base-uri 'self' https://*.mydomain.net;`
  csp += `form-action 'self' https://*.mydomain.net;`
  csp += `default-src 'self' https://*.mydomain.net 'nonce-${nonce}' data:;`

  return csp
}
```

**Reply by sophiekoonin** [reply]
@alex-r89 try splitting out your CSP to have a style-src rule, and pass the nonce into that. Does that help?

I know that Next also uses inline scripts and styles in dev mode. 
As far as I'm aware there are no inline scripts in production, so our `script-src 'self';` directive also has `'unsafe-inline'` in development mode only. If `NODE_ENV=production` we don't include this.

**Reply by Crelloc** [reply]
I don't find this to be the solution. Inline styles do not contain nonce. 
What version of Next are you using? Also, "res.setHeader('Content-Security-Policy', csp)" doesn't set the header, as in, I don't see explicitly see this header when analyzing pages.

**Reply by sophiekoonin** [reply]
I can't remember what version we're using off the top of my head but higher than v10.0.6 and lower than 10.1.0.

Setting the header in __document.js works for us, I don't know why it isn't working for you :( you're looking in the network tab of dev tools, right?


If you're looking for the nonce in dev tools, chrome hides it for security reasons so you won't see the actual nonce passed to the style tags - but you should see the `nonce` attribute on its own.

**Reply by alex-r89** [reply]
Hey @sophiekoonin thanks for the quick reply.

Splitting the style-src out (as well as the script-src) did not fix the issue :( I still am seeing the "Refused to execute" error in production. This is on next `v10.1.3`.

I also tried setting the `nonce` on the `NextScript` which also didnt work. Just for reference, this is my full `_document.js` file:


```js
import Document, { Html, Head, Main, NextScript } from 'next/document'
import { ServerStyleSheet } from 'styled-components'
import { randomBytes } from 'crypto'

function getCsp(nonce) {
  let csp = ``
  csp += `default-src 'self' https://*.mydomain.net https://*.amazonaws.com https://amazonaws.com 'nonce-${nonce}' data:;`
  csp += `style-src 'self' https://*.mydomain.net https://*.amazonaws.com https://amazonaws.com 'nonce-${nonce}' data:;`
  csp += `script-src 'self' https://*.mydomain.net https://*.amazonaws.com https://amazonaws.com 'nonce-${nonce}' data:;`

  return csp
}

let referrer = 'strict-origin'

export default class MyDocument extends Document {
  static async getInitialProps(ctx) {
    const nonce = randomBytes(8).toString('base64')
    const csp = getCsp(nonce)
    const sheet = new ServerStyleSheet()
    const originalRenderPage = ctx.renderPage

    const res = ctx?.res
    if (res != null) {
      res.setHeader('Content-Security-Policy', csp)
    }

    try {
      ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: (App) => (props) => sheet.collectStyles(<App {...props} />)
        })

      const initialProps = await Document.getInitialProps(ctx)

      return {
        ...initialProps,
        nonce,
        styles: (
          <>
            {initialProps.styles}
            {sheet.getStyleElement()}
          </>
        )
      }
    } finally {
      sheet.seal()
    }
  }

  render() {
    const { nonce } = this.props

    return (
      <Html lang='en' dir='ltr'>
        <Head nonce={nonce}>
          <meta name='referrer' content={referrer} />
        </Head>
        <body>
          <Main />
          <NextScript nonce={nonce} />
        </body>
      </Html>
    )
  }
}
```

**Reply by sophiekoonin** [reply]
I wouldn't know without testing it but I wonder if a recent version introduced a regression? 

I also don't know how it works with styled-components- I know this library uses inline styles. You'll need to pass your nonce to that somehow too

**Reply by alex-r89** [reply]
@sophiekoonin thanks for the help :) I am also getting the script Refused to execute error, so although the style error could be related to styled-components, Im wondering why the script error is being thrown. I'll keep playing around and see if I can get it working.

**Reply by sophiekoonin** [reply]
I just tested this in a brand new app with the latest version, and it doesn't work. That'd suggest why this post has got a bit of traction recently. I think one of the recent versions of Next must have broken this functionality.

This is the third or fourth problem we've had with an update breaking CSP (I've got an active issue for `next/image` not working properly with CSP, too) - this is getting tiresome.

**Reply by alex-r89** [reply]
@sophiekoonin thanks for investigating - yeah thats very annoying.

Ive changed my CSP headers to have `unsafe-eval' 'unsafe-inline'` for `script-src` and `'unsafe-inline'` for the `style-src`, which isnt ideal at all.

Which version of next were you using which was working?

**Reply by sophiekoonin** [reply]
Please don't use `unsafe-eval` in production! It's leaving you exposed to cross-site scripting vulnerabilities :( 

We're using `10.0.9` which seems to be working, so I'd start with rolling back to that.

**Reply by alex-r89** [reply]
Rolling back to `next v10.0.9` seems to have worked for me for the` script-src` errors (all `unsafe-`'s removed).

Styled components is still causing the `style-src` errors. It seems like passing a `nonce` in this same way isnt possible with styled-components and im unsure how to fix this at all - there doesn't seem to be a way to pass the `nonce` into styled-components. I may need to remove it all together, which is a pain.

But I can also attest that the current version of `nextjs` has broken the solution you provided @sophiekoonin


<details>
  <summary>Code</summary>
  
  ```js
import Document, { Html, Head, Main, NextScript } from 'next/document'
import { ServerStyleSheet } from 'styled-components'
import { randomBytes } from 'crypto'

const prod = process.env.NODE_ENV === 'production'

const inlineScript = (body, nonce) => (
  <script type='text/javascript' dangerouslySetInnerHTML={{ __html: body }} nonce={nonce} />
)

function getCsp(nonce) {
  let csp = ``
  if (prod) {
    csp += `default-src 'self' 'nonce-${nonce}' data: https://*.myDomain.net https://*.amazonaws.com https://amazonaws.com;`
    csp += `script-src 'self' 'self' 'nonce-${nonce}' data: https://*.myDomain.net https://*.amazonaws.com https://amazonaws.com;`
    csp += `style-src 'self' 'nonce-${nonce}' data: https://*.myDomain.net;`
  }

  return csp
}

export default class MyDocument extends Document {
  static async getInitialProps(ctx) {
    const sheet = new ServerStyleSheet()
    const originalRenderPage = ctx.renderPage
    const nonce = randomBytes(8).toString('base64')
    const csp = getCsp(nonce)
    const res = ctx?.res
    if (res != null) {
      res.setHeader('Content-Security-Policy', csp)
    }

    try {
      ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: (App) => (props) => sheet.collectStyles(<App {...props} />)
        })

      const initialProps = await Document.getInitialProps(ctx)

      return {
        ...initialProps,
        nonce,
        styles: (
          <>
            {initialProps.styles}
            {sheet.getStyleElement()}
          </>
        )
      }
    } finally {
      sheet.seal()
    }
  }

  render() {
    const { styles, nonce } = this.props

    return (
      <Html lang='en' dir='ltr'>
        <Head nonce={nonce}>
          {inlineScript(`window.__webpack_nonce__="${nonce}"`, nonce)}
          {styles}
        </Head>
        <body>
          <Main />
          <NextScript nonce={nonce} />
        </body>
      </Html>
    )
  }
}

  ```
  
</details>

EDIT: I actually dont think this is working - when I refresh the page a couple of times I eventually see a `Refused to execute inline script` error. Unsure why this is....back to the drawing board.

**Reply by redgumnfp** [reply]
As you use ctx?.res, I imagine you're using Server-Side Render at Runtime, so the below may possibly help.

Just wondering whether the issues you're having with the later versions is to do with getServerSideProps function being preferred now?
https://nextjs.org/docs/api-reference/data-fetching/getInitialProps

```
Recommended: getStaticProps or getServerSideProps.

If you're using Next.js 9.3 or newer, we recommend that you use getStaticProps or getServerSideProps instead of getInitialProps.

These new data fetching methods allow you to have a granular choice between static generation and server-side rendering. Learn more on the documentation for Pages and Data Fetching.
```

My sites use static export, so this doesn't help me too much - I've created hashes for any inline-scripts I'm using, however when it comes to inline-styles there are far too many to hash (the CSP would be massive)- they'd need to be in one file/style tag ideally. 

Regards,
James

**Reply by sophiekoonin** [reply]
@redgumnfp  These functions don't work in `_app.js` or `_document.js` yet: [Custom Document#Caveats](https://nextjs.org/docs/advanced-features/custom-document#caveats)

**Reply by redgumnfp** [reply]
Ah, good spot @sophiekoonin . I don't use the SSR feature, so wasn't sure if helpful or not.

I also tried doing it this way btw:
https://scotthelme.co.uk/csp-nonces-the-easy-way-with-cloudflare-workers/

This essentially acts like a man-in-the-middle to rewrite the document, so can work with static exports too.

Logically it's fine, but I have recently had issues with Cloudflare ontop of Vercel, so switched to Vercel directly the other week.

**Reply by sophiekoonin** [reply]
It all seems to be working on the latest version for me, though the key factor is that we've disabled Automatic Static Optimization via `getInitialProps` in `_app.js`. If that's not an option for you, the CSP-header solution may not work.

**Reply by sabirmgd** [reply]
Hey @sophiekoonin thanks for sharing this, this is been causing us issues, may I know which version is working for you, the script are working but not the styles for me, I'm using 10.0.5, its a server side rendering and attaching the nonce in the metadata:

```
export default class MyDocument extends Document {
  render(): React.ReactElement {
    const [csp, nonce] = generateCsp(); // function will create a nonce, its working for scripts but not for styles
    return (
      <Html lang="en">
        <Head nonce={nonce}>
          <meta property="csp-nonce" content={nonce} />
          <meta httpEquiv="Content-Security-Policy" content={csp} />
          <link rel="icon" href="/favicon.ico" />
          <link
            rel="preload"
            href={`${publicRuntimeConfig.BASE_PATH}/fonts/dmsans-regular-webfont.woff`}
            as="font"
            crossOrigin="anonymous"
          />
        </Head>
        <body>
          <Main />
          <NextScript nonce={nonce} />
        </body>
      </Html>
    );
  }
}
```

**Reply by sophiekoonin** [reply]
@sabirmgd if it's server-side rendering you could try setting the CSP on the response header as I did in my example above. not sure if that will help

**Reply by sabirmgd** [reply]
Thanks for your response, my `getInitialProps` doesn't get called in _document: so nonce is always undefined. I also have getInitialProps in _app

```
export default class MyDocument extends Document {
  static async getInitialProps(ctx) {
    console.log('Im being called');
    const initialProps = await Document.getInitialProps(ctx);
    const [csp, nonce] = generateCsp();

    const res = ctx?.res;
    if (res != null) {
      console.log("I'm in");
      res.setHeader('Content-Security-Policy', csp);
    }
    return {
      ...initialProps,
      nonce,
    };
  }

  render(): React.ReactElement {
    // const [csp, nonce] = generateCsp();
    const { nonce } = this.props;
    console.log(nonce);
    return (
      <Html lang="en">
        <Head nonce={nonce}>
```

**Reply by sabirmgd** [reply]
@sophiekoonin I'm getting: 
Property 'nonce' does not exist on type 'Readonly<RenderPageResult & { styles?: ReactElement<any, string | JSXElementConstructor<any>>[] | ReactFragment | undefined; } & { ...; }> & Readonly

**Reply by amedley** [reply]
This works: https://github.com/vercel/next.js/discussions/18451#discussioncomment-1975742

**Comment by Crelloc** [comment]
Can you confirm that nonce changes on every page refresh?

**Reply by sophiekoonin** [reply]
Yes, it does

**Comment by Tcheetox** [comment]
@sophiekoonin I tried your above solution to set nonce. There's something I don't quite understand though, it works perfectly fine in dev, but whenever I try the production environment on my machine (nothing fancy, npm run build && npm run start), I always end with: Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client.

I know it's not directly related, but I would appreciate any insights.

**Comment by dstroot** [comment]
Any updates here with Next 12?  What's the best way to have a good CSP without 'unsafe-inline' and 'unsafe-eval'

**Comment by Crelloc** [comment]
Check out this blog solution as it may help:

https://reesmorris.co.uk/blog/implementing-proper-csp-nextjs-styled-components

**Comment by amedley** [comment]
**SOLUTION FOR CLIENT RENDERED APPS**

Got this working with middleware and getInitialProps. You only have to SSR the `<Head>{...}</Head>` for this to work.

*pages/_middleware.js*
```
import {NextResponse} from 'next/server';
import {v4 as uuid} from 'uuid';

function csp(req, res) {
  const nonce = `nonce-${Buffer.from(uuid()).toString('base64')}`;
  const isProduction = process.env.NODE_ENV === 'production';
  const devScriptPolicy = ['unsafe-eval']; // NextJS uses react-refresh in dev
  res.headers.append('Content-Security-Policy', [
    ['default-src', 'self', nonce],
    ['script-src',  'self', nonce].concat(isProduction ? [] : devScriptPolicy),
    ['connect-src', 'self', nonce],
    ['img-src', 'self', nonce],
    ['style-src', 'self', nonce],
    ['base-uri',  'self', nonce],
    ['form-action', 'self', nonce],
  ].reduce((prev, [directive, ...policy]) => {
    return `${prev}${directive} ${policy.filter(Boolean).map(src => `'${src}'`).join(' ')};`
  }, ''));
}

export const middleware = (req) => {
  const res = NextResponse.next();
  csp(req, res);
  return res;
}
```

*pages/_app.js*
```
import Head from 'next/head';

const DisableSSR = ({children}) => {
  return (
    <div suppressHydrationWarning>
      {typeof window === 'undefined' ? null : children}
    </div>
  );
}

const Page = ({ Component, pageProps, nonce }) => {
  return (
    <div>
      <Head>
        <title>Create Next App</title>
        <meta name="description" content="Generated by create next app" />
        <meta property="csp-nonce" content={nonce} />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <DisableSSR>
        <Component {...pageProps} />
      </DisableSSR>
    </div>
  );
}

Page.getInitialProps = async ({ctx: {req, res}}) => {
  const csp = {};
  res.getHeaders()['content-security-policy']?.split(';').filter(Boolean).forEach(part => {
    const [directive, ...source] = part.split(' ');
    csp[directive] = source.map(s => s.slice(1, s.length - 1));
  });
  return {
    nonce: csp['default-src']?.find(s => s.startsWith('nonce-')).split('-')[1],
  };
};

export default Page;
```

**Comment by leerob** [comment]
Hey folks, wanted to swing back here with an update. After digging through many different issues and discussions, I've [**made a new page in the documentation**](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy) ([PR](https://github.com/vercel/next.js/pull/54601)) specifically for Content Security Policy and nonces. This docs page: 

- Explains how to generate a `nonce` with Middleware
- Shows how to consume the `nonce` in a route with `headers()`
- Shows a complete CSP without needing to use any `unsafe`
- Shows how to ignore the `nonce` Middleware from running on prefetches / static assets

Further, we've patched some bugs and made improvements to `nonce` handling in Next.js that will be available in the latest `canary` version (for those of you time traveling from the future, upgrade to Next.js 13.5). We also updated the `with-strict-csp` example in the `examples/` folder, which is backlinked from the [new documentation page](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy).

Really hope this helps out, thank you all 🙏 I'll be closing this discussion out. To continue the discussion, [please go here](https://github.com/vercel/next.js/discussions/54907).
